File: ASSIGNMENT10/python/Q.9.py
from collections import deque

class Graph:
    def __init__(self):
        self.adjacency_list = {}

    def add_edge(self, u, v):
        if u not in self.adjacency_list:
            self.adjacency_list[u] = []
        if v not in self.adjacency_list:
            self.adjacency_list[v] = []
        self.adjacency_list[u].append(v)
        self.adjacency_list[v].append(u)  # Uncomment this line for undirected graph

    def bfs(self, start):
        visited = set()      # Set to keep track of visited nodes
        queue = deque([start])  # Initialize the queue with the starting node
        bfs_order = []      # List to store the BFS traversal order

        while queue:
            vertex = queue.popleft()  # Dequeue a vertex from the queue
            if vertex not in visited:
                visited.add(vertex)   # Mark the vertex as visited
                bfs_order.append(vertex)  # Add it to the BFS order

                # Add all unvisited neighbors to the queue
                for neighbor in self.adjacency_list.get(vertex, []):
                    if neighbor not in visited:
                        queue.append(neighbor)

        return bfs_order

# Example Usage
if __name__ == "__main__":
    g = Graph()
    
    # Adding edges
    g.add_edge(0, 1)
    g.add_edge(0, 2)
    g.add_edge(1, 3)
    g.add_edge(1, 4)
    g.add_edge(2, 5)
    g.add_edge(2, 6)
    
    # Perform BFS starting from vertex 0
    bfs_result = g.bfs(0)

    # Print the result of BFS
    print("BFS Traversal starting from vertex 0:", bfs_result)

File: ASSIGNMENT10/c/Q.9.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Graph {
    int vertices;
    struct Node** adjLists;
};

struct Node* createNode(int v) {
    struct Node* newNode = malloc(sizeof(struct Node));
    newNode->data = v;
    newNode->next = NULL;
    return newNode;
}

struct Graph* createGraph(int vertices) {
    struct Graph* graph = malloc(sizeof(struct Graph));
    graph->vertices = vertices;

    graph->adjLists = malloc(vertices * sizeof(struct Node*));

    for (int i = 0; i < vertices; i++) {
        graph->adjLists[i] = NULL;
    }

    return graph;
}

void addEdge(struct Graph* graph, int src, int dest) {
    struct Node* newNode = createNode(dest);
    newNode->next = graph->adjLists[src];
    graph->adjLists[src] = newNode;
}

void bfs(struct Graph* graph, int startVertex) {
    int* visited = malloc(graph->vertices * sizeof(int));
    for (int i = 0; i < graph->vertices; i++) {
        visited[i] = 0;
    }

    struct Node* queue = NULL; // Initialize an empty queue

    visited[startVertex] = 1; // Mark the current node as visited
    printf("Visited %d\n", startVertex);

    struct Node* temp = graph->adjLists[startVertex];

    while (temp) {
        int adjVertex = temp->data;
        if (!visited[adjVertex]) {
            visited[adjVertex] = 1; // Mark it as visited
            printf("Visited %d\n", adjVertex);
        }
        temp = temp->next;
    }
}

int main() {
    struct Graph* graph = createGraph(5);
    addEdge(graph, 0, 1);
    addEdge(graph, 0, 4);
    addEdge(graph, 1, 2);
    addEdge(graph, 1, 3);
    addEdge(graph, 1, 4);
    addEdge(graph, 2, 3);
    addEdge(graph, 3, 4);

    printf("Breadth First Search starting from vertex 0:\n");
    bfs(graph, 0);
    return 0;
}
